### ISSUE TITLE
Filtering and Preparing the Data to finetune NLLB-200

### ISSUE THREAD

**Issue by ibtiRaj** [facebookresearch/stopes#15]
Hey,

I want to fine tune the NLLB 200 model, as instructed by the data [ReadMe](https://github.com/facebookresearch/fairseq/tree/nllb/examples/nllb/data) documentation, I have tried the filtering pipeline and got the output of populate_data_conf.py and compute_length_factors.py. 
But I don't know how to run prepare_data pipeline. Especially the parameter required by prepare_data.py : --data-config parameter, Could you provide an example? Thanks a lot.

**Comment by kauterry** [comment]
The README for `prepare_data` answers this. (https://github.com/facebookresearch/stopes/tree/main/stopes/pipelines/prepare_data) Here's an example config:

```
binarization_config:
  binarize_workers: 60
  max_examples_per_shard: 500000000
  random_seed: 0
  smallest_shard: 500000
executor_config:
  cluster: local
  log_folder: executor_logs
preprocessing_config:
  max_tokens: null
  moses_config:
    deescape_special_chars: false
    lowercase: false
    normalize_punctuation: true
    remove_non_printing_chars: false
    script_directory: <PATH_TO_FAIRSEQ_DIR>/fairseq-py/examples/nllb/modeling/preprocessing/moses
  preprocess_source: true
  preprocess_target: true
  sample_size: null
  tag_data: true
source_vocab_config:
  pretrained: null
  vocab_build_params:
    character_coverage: 0.99995
    model_type: bpe
    random_seed: 0
    sampled_data_size: 10000000
    sampling_temperature: 1.0
    shuffle_input_sentence: true
    use_joined_data: true
    vocab_size: 8000
target_vocab_config:
  pretrained: null
  vocab_build_params:
    character_coverage: 0.99995
    model_type: bpe
    random_seed: 0
    sampled_data_size: 10000000
    sampling_temperature: 1.0
    shuffle_input_sentence: true
    use_joined_data: true
    vocab_size: 8000
test_corpora:
  eng-ibo:
    values:
      flores_devtest:
        data_tag: null
        is_gzip: false
        num_lines: null
        source: <PATH>
        target: <PATH>
train_corpora:
  eng-ibo:
    values:
      public_bitext:
        data_tag: null
        is_gzip: false
        num_lines: null
        source: <PATH>
        target: <PATH>
train_mining_corpora: null
train_mmt_bt_corpora: null
train_smt_bt_corpora: null
valid_corpora:
  eng-ibo:
    values:
      flores_dev:
        data_tag: null
        is_gzip: false
        num_lines: null
        source: <PATH>
        target: <PATH>
```
The `data_path` format is detailed in the README, you need to organize your corpora files in a specific way for them to be read.

Make sure you download the moses scripts in your fairseq directory. The pipeline runs only if this script exists:
`examples/nllb/modeling/preprocessing/moses/clean-corpus-n.perl` (https://github.com/facebookresearch/stopes/blob/main/stopes/pipelines/prepare_data/encode_and_binarize.py#L78)

**Comment by kauterry** [comment]
You need to run the filtering pipeline to filter out data based on the following heuristics: length, deduplication, LASER margin score threshold, LID score thresholds, toxicity. It's not sufficient to just get the output of `populate_data_conf.py` and `compute_length_factors.py`, the output of these scripts are then passed into the filtering pipeline. This is detailed in the README here: https://github.com/facebookresearch/stopes/tree/main/stopes/pipelines/filtering

After filtering, you build a vocabulary (SentencePiece model) and then encode and binarize your data which can then be fed into fairseq for training. The filtered datasets (output of filtering pipeline) should be fed into the `prepare_data` pipeline.

**Comment by ibtiRaj** [comment]
@kauterry Thank you very much for your detailed answer. I tried to do what you said, first I run the filter script,  then I created a config.yaml file as the file you shared with me:
![config](https://user-images.githubusercontent.com/50990415/198672412-fbb5f282-4a17-4e02-bd12-289f3954ea3d.png)
But when I run the prepare_data.py I got an error:
![error](https://user-images.githubusercontent.com/50990415/198673030-9493f6e6-40fd-4654-905c-a204bc86824c.png)

**Comment by ibtiRaj** [comment]
> You need to run the filtering pipeline to filter out data based on the following heuristics: length, deduplication, LASER margin score threshold, LID score thresholds, toxicity. It's not sufficient to just get the output of `populate_data_conf.py` and `compute_length_factors.py`, the output of these scripts are then passed into the filtering pipeline. This is detailed in the README here: https://github.com/facebookresearch/stopes/tree/main/stopes/pipelines/filtering
> 
> After filtering, you build a vocabulary (SentencePiece model) and then encode and binarize your data which can then be fed into fairseq for training. The filtered datasets (output of filtering pipeline) should be fed into the `prepare_data` pipeline.

You said that output of filtering pipeline should be fed into the prepare_data pipeline. but I don't know where and how, such as the parameters of prepare_data.py are:
- data_config: config.yaml,  the config file that we made
- Data_path: the path of my data
- output_dir

I'm sorry if I misunderstood and I thank you.

**Comment by kauterry** [comment]
You shouldn't include `perl` in your moses script directory.

**Comment by ibtiRaj** [comment]
Yes, you're right thanks. But still the same error: "empty training data"

**Comment by kauterry** [comment]
Can you `ls` your `script_moses` directory and show me the output?

**Comment by ibtiRaj** [comment]
![ls](https://user-images.githubusercontent.com/50990415/198681701-d595c822-fa9e-466c-b7a3-f24e32bd1c2b.png)

**Comment by kauterry** [comment]
This file needs to exist at this exact path: `moses_config.script_directory/normalize-punctuation.perl`

**Comment by ibtiRaj** [comment]
I've tried to do what you said:
![image](https://user-images.githubusercontent.com/50990415/199035394-48a98df4-0102-434e-8ac9-e3bb030d892e.png)
But I still get the same error. So I tried to trace the error and see what is causing the problem.
I think the problem is that when prepare_data.py calls the validate_data_config function on line 297, it seems that this function returns empty folders. Do you think this is really the cause of the problem?

![image](https://user-images.githubusercontent.com/50990415/199037177-db48f604-1894-4413-bbfa-94eb1c2649a3.png)

![image](https://user-images.githubusercontent.com/50990415/199037340-05e01eb7-4045-4d9c-8aed-5ecf2e902056.png)

![image](https://user-images.githubusercontent.com/50990415/199037565-786a8da2-2166-4e8d-b378-21a23a8d34a0.png)

**Comment by ibtiRaj** [comment]
@kauterry  Thank you very much, I have solved the problem and finished the first and second step: filtering and data preparation.

**Comment by ibtiRaj** [comment]
hey @kauterry, I tried to fine tune NLLB model with my prepared data, but I am getting errors about missing files in the data directory. So I need your help to know if the result of the Prepare Data step looks like the following directory:
![image](https://user-images.githubusercontent.com/50990415/204840777-b3bbff71-aa9a-40f1-94fb-5382f1acc8e3.png)

The data_bin directory includes a shard000 directory that contains the following file:
![image](https://user-images.githubusercontent.com/50990415/204842303-d60ea9dc-4f40-4a39-a640-6b66dba2cf48.png)

The error I get when I try to fine tune the NLLB model is this:
![image](https://user-images.githubusercontent.com/50990415/204843132-daa4f6f2-46b8-4f33-ab90-15fc6a3a7c78.png)

Thank you in advance
